### PR TITLE
Fix latency tracing; Support batch_size > mini_batch_size; Support gradient accumulation

### PR DESCRIPTION
**Pull request by liujch1998**
WANDB log showing that things are working fine (especially the rollouts) with B=64, MB=32, grad_acc=2
https://wandb.ai/liujch1998/n-Tulu-PPO-Jax/runs/train_v2.2.1_v2.2_debugging_t256_m1x32x8_b64_mb32_g2--20fdca63358140e6b618373537bda962

Latency tracing: `ppo_total` is everything for a batch, it is equal to `ppo_rollout` plus everything else (forward, backward, optimizer step ...)
Screenshot 2024-04-11 at 10 46 29

### CODE REVIEW

File: EasyLM/models/llama/llama_train_ppo.py [hunk 1]
```diff
@@ -175,26 +176,22 @@ def compute_advantages(values, rewards, mask):
...
+def ppo_rollout(
+    policy_train_state,
+    policy_model,
     rng, batch,
 ):
     rng_generator = JaxRNG(rng)
-    batch = with_sharding_constraint(batch, PS(('dp', 'fsdp')))
+    batch = with_sharding_constraint(batch, PS(('dp', 'fsdp'))) # dim 0 is sharded across dp and fsdp axes, dim 1 is not sharded
+
+    pad_token_id = 0
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: Is there a way we can avoid hardcoding token ids like this?

File: EasyLM/models/llama/llama_train_ppo.py [hunk 2]
```diff
@@ -82,6 +82,7 @@
     generate_only=False,
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py:84]:
```suggestion
```

we don't need this flag anymore, I think?

**Review by hamishivi — APPROVED** [review]
LGTM, assuming you have tested and everything works.

**Comment by liujch1998** [comment]
Deprecating this PR, moving things to <https://github.com/hamishivi/EasyLM/pull/19>

**Pull request closed without merging**